docs(mempool/lanes): Add technical design section to ADR (#3242)
cometbft/cometbft: docs/references/architecture/adr-113-mempool-lanes.md

Before jumping into the design of the proposal, we define more formally the properties supported by
the current implementation of the mempool. Then we state what properties the new mempool should
offer to guarantee the desired QoS. The following definitions are common to all properties.

When attempting to add an incoming transaction to the mempool, the node first checks that it is not
already in the cache before checking its validity with the application. 

:memo: _Definition_: We say that a node receives a transaction `tx` _for the first time_ when the
node receives `tx` and `tx` is not in the cache.

By this definition, it is possible that a node receives a transaction "for the first time", then
gets the transaction evicted from the cache, and at a later time receives it "for the first time"
again. The cache implements a Least-Recently Used (LRU) policy for removing entries when the
cache is full.

:memo: _Definition_: Given any two different transactions `tx1` and `tx2`, in a given node, we say that:
1. `tx1` is *validated before* `tx2`, when `tx1` and `tx2` are received for the first time, and `tx1`
is validated against the application (via `CheckTx`) before `tx2`,

> 1. `tx1` is *rechecked before* `tx2`, when `tx` and `tx2` are in the mempool and `tx1` is

Reviewer: ```suggestion
1. `tx1` is *rechecked before* `tx2`, when `tx1` and `tx2` are in the mempool and `tx1` is
```

> re-validated (rechecked via `CheckTx`) before `tx2`,

Reviewer: What if `tx1` and `tx2` are re-checked in different heights? Does that case fit into this definition, or we're leaving such a case out on purpose?

Author: If both transactions are in the mempool, that means that they are rechecked at the same height. I'll clarify in the text.

1. `tx1` is *reaped before* `tx2`, when `tx1` is reaped from the mempool to be included in a block
  proposal before `tx2`,
1. `tx1` is *disseminated before* `tx2`, when `tx1` is sent to a given peer before `tx2`.

In 4, note that in the current implementation there is one dissemination routine per peer, so it

> could happen that `tx2` is sent to a peer before `tx1` is sent to a different peer.

Reviewer: ```suggestion
could happen that `tx2` is sent to a peer before `tx1` is sent to a different peer.
Hence the importance of expression "to a given peer" in that definition.
```

### Current mempool

As stated above, the current mempool offers a best-effort FIFO ordering of transactions. We state
this property as follows.

:parking: _Property_ **FIFO ordering of transactions**: We say that the mempool makes a best effort

> in maintaining the FIFO ordering of transactions when transactions are validated, rechecked, reaped,

Reviewer: Excellent!

characteristics as defined by the application.

A transaction may only have one class. If it is not assigned any specific class, it will be assigned
a *default class*, which is a special class always present in any set of classes. This is analogous
to the _native VLAN_ for untagged traffic in an 802.1Q network. Because no transaction can belong to

> two or more classes, transaction classes form disjoint sets, that is, the intersection between

Reviewer: ```suggestion
two or more classes, transaction classes form disjoint sets, that is, the intersection between any two
```

characteristics as defined by the application.

A transaction may only have one class. If it is not assigned any specific class, it will be assigned
a *default class*, which is a special class always present in any set of classes. This is analogous
to the _native VLAN_ for untagged traffic in an 802.1Q network. Because no transaction can belong to
two or more classes, transaction classes form disjoint sets, that is, the intersection between any two
classes is empty. Also, all transactions in the mempool are the union of the transactions in all
classes.

:memo: _Definition_: Each class has a *priority* and two classes cannot have the same priority.
Therefore all classes can be ordered by priority.

> When a transaction is received for the first time and validated via `CheckTx`, the application MUST

Reviewer: But, we say somewhere that, if the add doesn't return a class, then the TX belongs to the default class?

Author: True. I fixed it.

We want also to keep the FIFO ordering within each class (for the time being):

:parking: _Property_ **FIFO ordering per class**: For transactions within the same class, the
mempool will maintain a FIFO order within the class when transactions are validated, rechecked,
reaped, and disseminated.

Given any two different transactions `tx1` and `tx2` belonging to the same class, if the mempool
receives `tx1` before receiving `tx2`, then:
- `tx1` will be validated and recheck against the application (via `CheckTx`) before `tx2`, and
- `tx1` will be reaped and disseminated before `tx2`.

As a consequence, given that classes of transactions have a sequential ordering, and that classes do
not have elements in common, we can state the following property:

:parking: _Property_ **Partial ordering of all transactions**: The set of all the transactions in
the mempool, regardless of their classes, will have a *partial order*.

> This means that some pairs of transactions are comparable and, thus, have and order, while others

Reviewer: ```suggestion
This means that some pairs of transactions are comparable and, thus, have an order, while others
```

For broadcasting transactions from multiple lanes, we have considered two possible approaches:
1. Reserve $N$ p2p channels for use by the mempool. P2P channels have priorities that we can reuse
   as lane priorities. There are a maximum of 256 P2P channels, thus limiting the number of lanes.
2. Continue using the current P2P channel for disseminating transactions and implement logic within
   the mempool to select the order of transactions to put in the channel. This option theoretically
   allows for an unlimited number of lanes, constrained only by the nodes’ capacity to store the
   lane data structures.

We choose the second option for its flexibility, allowing us to start with a simple scheduling
algorithm that can be refined over time. The desired algorithm must satisfy the properties
"Priorities between classes" and "FIFO ordering per class". This requires supporting selection by
_weight_, ensuring each lane gets a fraction of the P2P channel capacity proportional to its
priority. Moreover, it should be _fair_, to prevent starvation of low-priority lanes. Given the
extensive research on scheduling algorithms in operating systems and networking, we propose for the

> MVP implementing a variant of the [Weighted Round Robin (WRR)](wrr) algorithm, which meets these

Reviewer: In QoS for IP networks there is [WRED](https://en.wikipedia.org/wiki/Weighted_random_early_detection). I need check if that would be relevant here.

Reviewer 2: The link is broken.

Reviewer 2: Again, is this what the p2p channel does?

Reviewer: > The link is broken.

Which link do you refer to? The one on the text, or the one in my comment (both seem to be working for me)

> Again, is this what the p2p channel does?

Yeah, I had the same question when I read this part of Thu. Is priority in p2p channels implementing `wrr` or something else?

Author: The link on WRR was broken and then I fixed it. I added a section called Prioritization Logic with a description of the algorithm in P2P channels.

requirements and is straightforward to implement.

> Before modifying the dessemination logic, we need to refactor the current implementation and the

Reviewer: ```suggestion
Before modifying the dissemination logic, we need to refactor the current implementation and the
```

`Mempool` interface to clearly separate the broadcast goroutine in the mempool reactor from
`CListMempool` that includes the mempool data structures. `CListMempool` provides two methods used
by the broadcast code, `TxsWaitChan() <-chan struct{}` and `TxsFront() *clist.CElement`, which are
just wrappers around the methods `WaitChan` and `Front` of the `CList` implementation. In
particular, `TxsFront` is leaking implementation details outside the `Mempool` interface.

### Checking received transactions

Transactions are transmitted without lane information because peers cannot be trusted to send the
correct data. A node may take advantage of the network by sending lower-priority transactions before
higher-priority ones. Although the receiving node could easily verify the priority of a transaction
when it calls `CheckTx`, it cannot detect if a peer is sending transactions out of order over a

> single P2P channel. For the moment, we leave out of the MVP any mechanism to detect and possible

Reviewer: ```suggestion
single P2P channel. For the moment, we leave out of the MVP any mechanism to detect and possibly
```

Another drawback is that this option imposes a limit to the number of P2P channels. Channels use a
byte as an ID, and the current distribution among all reactors goes up to channel `0x61`. For
example, the current mempool’s P2P channel ID is `0x30`, which would serve as the default lane. We
could reserve a range of channels for the mempool, such as starting from channel ID `0x80` and above
(all channels with the most significant byte set to 1). This would provide a maximum of 128 lanes,
which should suffice for most users.

Nodes would also need to agree on the channel assignments during the P2P handshake. Currently, one
of the conditions for the handshake to succeed is that there must exist an intersection of P2P
channels. Since lanes are defined in the application logic, the nodes only need to agree on the
application version, as it already happens in the current implementation.

### Duality lane/priority

The duality lane/priority could introduce a powerful indirection. The app could just define the lane
of a transaction in `CheckTx`, but the priority of the lane itself could be configured (and

> fine-tuned) elsewhere. For example, by the app itself or by node operators. The proposed design does

Reviewer: ```suggestion
fine-tuned) elsewhere. For example, by the app itself or by node operators. The proposed design for the MVP does
```

not support this pattern.

### Custom configuration per lane

A straightforward, future improvement that we leave for after the MVP is to allow customized
configuration of the lanes instead of sharing the current mempool configuration among lanes. The
application would need to define new configuration values per lane and pass them to CometBFT during
the handshake.

### Where to define lanes and priorities

We have considered two alternative approaches for _where_ to configure lanes and priorities:
1. In `config.toml` or `app.toml`. We have discarded this option as it does not make sense for

>    different nodes to have different lane configurations.

Reviewer: ```suggestion
   different nodes to have different lane configurations. The properties defined in the specification above
   are end-to-end, and so, the lane configuration has to be consistent across the network.
```

1. In `ConsensusParams`. There are several disadvantages with this approach. If we allow changing
   lane information via `ConsensusParams`, the mempool would need to update lanes dynamically. The
   updating process would be very complex and cumbersome, and not really appealing for an MVP. Two

>    governance proposals would be required to pass to update the lane definitions. A first proposal

Reviewer: ```suggestion
   governance proposals would be required to pass to update (a) the lane definitions, and (b) a new version of the app with adapted classification logic.
   A first proposal
```

   would be required for upgrading the application, because the lane classification logic (thus the
   application's code) needs to know the lane configuration beforehand. And a second proposal would

>    be needed for upgrading the lanes via `ConsensusParams`.

Reviewer: ```suggestion
   be needed for upgrading the lanes via `ConsensusParams`.
   While it is true that SDK applications could pass a governance proposal with both elements together,
   it would be something to _always_ do, and it is not clear what the situation would be for non-SDK applications.
```

  
   Also, it is not clear in which order the proposals should apply. The community should be careful
   not to break performance between the passing of both proposals. The `gov` module could be
   modified to allow the two changes to be shipped in the same gov proposal, but this does not seem
   a feasible solution.

Moreover, these two alternatives have a common problem which is how to deal with nodes that are
late, possibly having lane definitions that do not match with those of nodes at the latest heights.

## Consequences

TODO

### Positive

- Application developers will be able to better predict when transactions will be disseminated and
  processed to be included in a block. This has direct impact on block creation and transaction
  latency.
- The mempool will be able to offer Quality of Service guarantees, which does not exist in the
  current implementation. 
- Lanes will preserve the "FIFO ordering of transactions" property within the same class (with a
  best effort approach, as the current implementation).

> 

Reviewer: ```suggestion
- Applications that are unaware of this feature, and therefore not classifying transactions in `CheckTX`
  will observe the same behavior from the mempool as the current implementation.  
```

### Negative

- The total FIFO ordering of transaction dissemination and block creation may be broken when
  multiple lanes are used. Since FIFO ordering is important within the same class of transactions,
  we expect this will not be a real problem.

> 

Reviewer: ```suggestion
- Increased complexity in the logic of `CheckTx` (ante handlers) in order to classify transactions,
  with a possibility of introducing bugs in the classification logic.
```

### Neutral

- Lanes are optional. Current applications do not need to make any change to their code. Future
  applications will not be forced to use the lanes feature.
- The proposed prioritization logic (WRR) for the dissemination algorithm is fair, so low-priority

>   transactions will not get stuck in the mempool long periods of time.

Reviewer: ```suggestion
  transactions will not get stuck in the mempool for long periods of time.
```

## Detailed Design

> This sections describes the architectural changes needed to implement an MVP of lanes in the

Reviewer: ```suggestion
This section describes the architectural changes needed to implement an MVP of lanes in the
```

mempool. The following is a summary of the key design decisions:

> - [[1](#lanes-definition)] The list of lanes and their corresponding priorities

Reviewer: ```suggestion
- [[Lanes definition](#lanes-definition)] The list of lanes and their corresponding priorities
```

  will be hardcoded in the application logic.

> - [[2](#internal-data-structures)] There will be one concurrent list (CList) data structure per

Reviewer: ```suggestion
- [[Internal data structures](#internal-data-structures)] There will be one concurrent list (CList) data structure per
```

  lane.

> - [[3](#configuration)] All lanes will share the same mempool configuration.

Reviewer: ```suggestion
- [[Configuration](#configuration)] All lanes will share the same mempool configuration.
```

> - [[4](#adding-transactions-to-the-mempool)] When validating a transaction via CheckTx, the

Reviewer: ```suggestion
- [[Adding transactions to the mempool](#adding-transactions-to-the-mempool)] When validating a transaction via CheckTx, the
```

  application will optionally return a lane for the transaction.

> - [[5](#reaping-transactions-for-creating-blocks)] Transactions will be reaped from higher-priority

Reviewer: ```suggestion
- [[Reaping transactions for creating blocks](#reaping-transactions-for-creating-blocks)] Transactions will be reaped from higher-priority
```

>   lanes first, preserving the FIFO ordering.

Reviewer: ```suggestion
  lanes first, preserving intra-lane FIFO ordering.
```

> - [[6](#transaction-dissemination)] We will continue to use the current P2P channel for

Reviewer: ```suggestion
- [[Transaction dissemination](#transaction-dissemination)] We will continue to use the current P2P channel for
```

   disseminating transactions, and we will implement in the mempool the logic for selecting the
   order in which to send transactions.

### Lanes definition

The list of lanes and their corresponding priorities will be hardcoded in the application logic. A
priority is a value of type `uint32`, with 0 being a reserved value (see below). The application
also needs to define which of the lanes it defines is the default lane. 

To obtain the lane information from the application, we need to extend the ABCI `Info` response to

> include the following fields. These fields need to be filled by the application only in case it

Reviewer: ```suggestion
include the following fields.
Remember CometBFT calls `Info` when the node is starting up (or recovering) as part of CometBFT handshake with the application.
These fields need to be filled by the application only in case it
```

wants to implement lanes.
```protobuf
message InfoResponse {
  ...

>   repeated uint32 lanes = 6;

Reviewer: `lane_priorities` maybe?

Author: Changed to `lane_priorities` and `default_lane_priority`.

  uint32 default_lane = 7;
}
```
Internally, the application may use `string`s to name lanes, and then map those names to priorities.
The mempool does not care about the names, only about the priorities. That is why the lane
information returned by the application only contains priorities.

> The lowest priority a lane may have is 1. The value 0 is reserved for two cases: when there are no
> priorities and for `CheckTx` responses of invalid transactions.

Reviewer: ```suggestion
The lowest priority a lane may have is 1. The value 0 is reserved for two cases: when the application does not classify the transaction (i.e. no priority returned) and for `CheckTx` responses of invalid transactions.
```

On receiving the information from the app, CometBFT will validate that:

> - `lanes` has no duplicates (values in `lanes` don't need to be sorted),

Reviewer: Would it make our life easier to require that the lanes are sorted (to simplify the validation code)?

Author: Not really. It's easy to sort the list and we wouldn't need to trust the app to do it.

> - `default_value` is in `lanes` (the default lane is not necessarily the lane with the lowest

Reviewer: ```suggestion
- `default_lane` is in `lanes` (the default lane is not necessarily the lane with the lowest
```

  priority), and
- the list `lanes` is empty if and only if `default_value` is 0.

> ### Initialization

Reviewer: Taking care of this...

> ### Initialization

Reviewer: Done

Currently, querying the app for `Info` happens during the handshake between CometBFT and the app,
during the node initialization, and only when state sync is not enabled. The `Handshaker` first
sends an `Info` request to fetch the app information, and then replays any stored block needed to
sync CometBFT with the app. The lane information is needed regardless of whether state sync is
enabled, so one option is to query the app information outside of the `Handshaker`.

In this proposed approach, updating the lane definitions will require a single governance proposal
for updating the software. CometBFT will not need to deal with dynamic lane changes: it will just
need to set up the lanes when starting up (whether afresh or in recovery mode).

Different nodes also need to agree on the lanes they use. When a node connects to a peer, they both
perform a handshake to agree on some basic information (see `DefaultNodeInfo`), including the
version of the application they are executing. Since the application includes the lane definitions,
both nodes will also agree on the lanes they implement. Therefore we don't need to modify the P2P
handshake.

> 

Reviewer: We could add a paragraph here to explain compatibility with CosmoVisor-like approaches.

Author: Yes. That's important and I forgot.

Reviewer: Will add a paragraph

Reviewer: Working on it now...

Reviewer: Done. Please take a look

Author: Thanks!

```golang
type Lane uint32
```

Currently, the `CListMempool` data structure has two fields to store and access transactions:
```golang
txs    *clist.CList // Concurrent list of mempool entries.
txsMap sync.Map     // Map of type TxKey -> *clist.CElement, for quick access to elements in txs.
```

With the introduction of lanes, the main change will be to divide the `CList` data structure into
$N$ `Clist`s, one per lane. `CListMempool` will have the following fields:
```golang
lanes   map[Lane]*clist.CList
txsMap  sync.Map // Map of type TxKey -> *clist.CElement, for quick access to elements in lanes.
txLanes sync.Map // Map of type TxKey -> Lane, for quick access to the lane corresponding to a tx.

// Fixed variables set during initialization.
defaultLane Lane
sortedLanes []Lane // Lanes sorted by priority
```
The auxiliary fields `txsMap` and `txLanes` are, respectively, for direct access to the mempool
entries, and for direct access to the lane of a given transaction.

If the application does not implement lanes (that is, it responds with empty values in
`InfoResponse`), then `defaultLane` will be set to 1, and `lanes` will have only one entry for the

> default lane.

Reviewer: ```suggestion
default lane. In this case, the new mempool's behaviour will be equivalent to that of the current mempool.
```

`CListMempool` also contains the cache, which is only needed before transactions have a lane
assigned. Since the cache is independent of the lanes, we do not need to modify it.

### Configuration

For an MVP, all lanes can share the same mempool configuration. In this scenario, all lanes will be
capped by the total mempool capacities as currently defined in the configuration. Namely, these are:
- `Size`, the total number of transactions allowed in the mempool, 
- `MaxTxsBytes`, the maximum total number of bytes of the mempool, and
- `MaxTxBytes`, the maximum size in bytes of a single transaction accepted into the mempool.

> With lanes, the total size of the mempool will be the sum of the sizes of all lanes.

Reviewer: Do we want to divide the total size evenly across lanes? (or are we gonna expunge a TX from a lower lane when a full mempool receives a TX?)

Author: I'd say it's not needed for the MVP but it could be an interesting improvement for the future. Equal sizes for all lanes may not make sense when lanes are defined by transaction size. For example, a lane for only very big txs. Probably we won't see this issue when we add custom configurations per lane.

Additionally, the `Recheck` and `Broadcast` flags will apply to all lanes or to none. Remember that,

> with `PrepareProposal`, it becomes _always_ mandatory to recheck remaining transactions in the
> mempool, so there is no point in disabling `Recheck` per lane.

Reviewer: ```suggestion
if `PrepareProposal`'s app logic can ever add a new transaction, it becomes _always_ mandatory to recheck remaining transactions in the
mempool, so there is no point in disabling `Recheck` per lane.
```

### Adding transactions to the mempool

When validating a transaction received for the first time with `CheckTx`, the application will
optionally return its lane in the response.
```protobuf
message CheckTxResponse {
  ...
  uint32 lane = 12;
}
```
The callback that handles the first-time CheckTx response will append the new mempool entry to the
corresponding `CList`, namely `lanes[lane]`, and update the other auxiliary variables accordingly.
If `lane` is 0, it means that the application did not set any lane in the response message, so the
transaction will be assigned to the default lane.

> ### Removing transactions from the mempool

Reviewer: Should this also be present in the list at the beginning of the section?

Author: I didn't add it because there's nothing important about removing transactions.

- [[Adding transactions to the mempool](#adding-transactions-to-the-mempool)] When validating a transaction via CheckTx, the
  application will optionally return a lane for the transaction.
- [[Reaping transactions for creating blocks](#reaping-transactions-for-creating-blocks)] Transactions will be reaped from higher-priority
  lanes first, preserving intra-lane FIFO ordering.
- [[Transaction dissemination](#transaction-dissemination)] We will continue to use the current P2P channel for
   disseminating transactions, and we will implement in the mempool the logic for selecting the
   order in which to send transactions.

### Lanes definition

The list of lanes and their corresponding priorities will be hardcoded in the application logic. A
priority is a value of type `uint32`, with 0 being a reserved value (see below). The application
also needs to define which of the lanes it defines is the default lane. 

To obtain the lane information from the application, we need to extend the ABCI `Info` response to
include the following fields. These fields need to be filled by the application only in case it
wants to implement lanes.
```protobuf

> message InfoResponse {

Reviewer 2: The one from lines 327 to 331.

By the conditions on lines 342 and 343.

  ...
  repeated uint32 lanes = 6;
  uint32 default_lane = 7;
}
```
Internally, the application may use `string`s to name lanes, and then map those names to priorities.
The mempool does not care about the names, only about the priorities. That is why the lane
information returned by the application only contains priorities.

The lowest priority a lane may have is 1. The value 0 is reserved for two cases: when the application does not classify the transaction (i.e. no priority returned) and for `CheckTx` responses of invalid transactions.

On receiving the information from the app, CometBFT will validate that:
- `lanes` has no duplicates (values in `lanes` don't need to be sorted),
- `default_lane` is in `lanes` (the default lane is not necessarily the lane with the lowest
  priority), and
- the list `lanes` is empty if and only if `default_value` is 0.

### Initialization

> Currently, querying the app for `Info` happens during the handshake between CometBFT and the app,

Author: Not specifically this scenario but section "Validating lanes of received transactions" is about what to do when a node decides to broadcast txs in whatever order.

```

With the introduction of lanes, the main change will be to divide the `CList` data structure into
$N$ `Clist`s, one per lane. `CListMempool` will have the following fields:
```golang
lanes   map[Lane]*clist.CList
txsMap  sync.Map // Map of type TxKey -> *clist.CElement, for quick access to elements in lanes.
txLanes sync.Map // Map of type TxKey -> Lane, for quick access to the lane corresponding to a tx.

// Fixed variables set during initialization.
defaultLane Lane
sortedLanes []Lane // Lanes sorted by priority
```
The auxiliary fields `txsMap` and `txLanes` are, respectively, for direct access to the mempool
entries, and for direct access to the lane of a given transaction.

If the application does not implement lanes (that is, it responds with empty values in
`InfoResponse`), then `defaultLane` will be set to 1, and `lanes` will have only one entry for the
default lane. In this case, the new mempool's behaviour will be equivalent to that of the current mempool.

`CListMempool` also contains the cache, which is only needed before transactions have a lane
assigned. Since the cache is independent of the lanes, we do not need to modify it.

### Configuration

> For an MVP, all lanes can share the same mempool configuration. In this scenario, all lanes will be

Author: You're right. I rephrased it.

> For an MVP, all lanes can share the same mempool configuration. In this scenario, all lanes will be

Author: Agree, this is not clear or correct. I rephrased it.

capped by the total mempool capacities as currently defined in the configuration. Namely, these are:
- `Size`, the total number of transactions allowed in the mempool, 
- `MaxTxsBytes`, the maximum total number of bytes of the mempool, and
- `MaxTxBytes`, the maximum size in bytes of a single transaction accepted into the mempool.

With lanes, the total size of the mempool will be the sum of the sizes of all lanes.

Additionally, the `Recheck` and `Broadcast` flags will apply to all lanes or to none. Remember that,
if `PrepareProposal`'s app logic can ever add a new transaction, it becomes _always_ mandatory to recheck remaining transactions in the
mempool, so there is no point in disabling `Recheck` per lane.

### Adding transactions to the mempool

When validating a transaction received for the first time with `CheckTx`, the application will
optionally return its lane in the response.
```protobuf
message CheckTxResponse {
  ...
  uint32 lane = 12;
}
```
The callback that handles the first-time CheckTx response will append the new mempool entry to the
corresponding `CList`, namely `lanes[lane]`, and update the other auxiliary variables accordingly.

> If `lane` is 0, it means that the application did not set any lane in the response message, so the

Author: The application implicitly returns 0 when it ignores lanes in the CheckTx response. We cannot force the application to return a specific value when they don't want to implement lanes, so the mempool needs to take care of assigning the default lane.

transaction will be assigned to the default lane.

### Removing transactions from the mempool

A transaction may be removed in two scenarios: when updating a list of committed transactions, and
during rechecking if the transaction is reassessed as invalid. In either case, the first step is to
identify the lane the transaction belongs to by accessing the `txLanes` map. Then, we remove the
entry from the CList corresponding to its lane and update the auxiliary variables accordingly.

Since the broadcast goroutines are constantly reading the list of transactions to disseminate them,
it's important to prioritize the removal of transactions from high-priority lanes first.

When updating the mempool, there is potential for a slight optimization by removing transactions
from different lanes in parallel. To achieve this, we would first need to preprocess the list of
transactions to determine the lane of each transaction. However, this optimization has minimal
impact if the committed block contains few transactions. Therefore, we decided to exclude it from
the MVP.

> ### Reaping transactions for creating blocks

Reviewer 2: Any discussion regarding starvation?

Namely, no guarantee that a tx in a low priority lane will eventually be included in the output?

I mentioned that to you guys at some point, and I still think that the way the `p2p/MConnection` does that is probably more adequate. I think that considering this approach should at least be discussed here, as it is an ADR.

Author: It was not mentioned so I added it. In essence it's the same as for tx dissemination. I agree we should discuss what P2P does and compare it to the proposed scheduling algorithm.

Author: I added a section Prioritization logic that includes the current scheduling algorithm in the P2P reactor.

Before modifying the dissemination logic, we need to refactor the current implementation and the
`Mempool` interface to clearly separate the broadcast goroutine in the mempool reactor from
`CListMempool` that includes the mempool data structures. `CListMempool` provides two methods used
by the broadcast code, `TxsWaitChan() <-chan struct{}` and `TxsFront() *clist.CElement`, which are
just wrappers around the methods `WaitChan` and `Front` of the `CList` implementation. In
particular, `TxsFront` is leaking implementation details outside the `Mempool` interface.

### Checking received transactions

Transactions are transmitted without lane information because peers cannot be trusted to send the
correct data. A node may take advantage of the network by sending lower-priority transactions before
higher-priority ones. Although the receiving node could easily verify the priority of a transaction
when it calls `CheckTx`, it cannot detect if a peer is sending transactions out of order over a
single P2P channel. For the moment, we leave out of the MVP any mechanism to detect and possibly
punish nodes for this kind of behaviour.

## Alternative designs

### One CList for all lanes

> We briefly considered sharing one CList for all lanes, changing the internal logic of CList to

Reviewer 2: This is the priority mempool, in other words.

Author: I wouldn't say so because in the priority mempool every tx has a priority. The idea of the "one CList for all lanes" solution was something like adding a lane field to each entry and then other auxiliary data structures to support traversing the entries per lane.

Reviewer 2: So it is not one clist for all lanes, but one clist and multiple clists, one per lane, using the same entries as the one big clist.

In other words, a linked-list that contains other abstract linked-lists inside it.

accommodate lane requirements. However, this design significantly increases code complexity,
particularly in the transaction dissemination logic.

> ### One P2P channel per lane

Reviewer 2: Also, we would need a different routine per lane/channel, otherwise there is no real difference between the approaches.

Also consider that the assignment of channels is **static**.

simply append new transactions to their respective channels and allow the P2P layer to manage the
order of transmission. We decided against this option mainly because the prioritization logic cannot
be easily modified without altering the P2P code, potentially affecting other non-mempool channels.

Another drawback is that this option imposes a limit to the number of P2P channels. Channels use a
byte as an ID, and the current distribution among all reactors goes up to channel `0x61`. For
example, the current mempool’s P2P channel ID is `0x30`, which would serve as the default lane. We
could reserve a range of channels for the mempool, such as starting from channel ID `0x80` and above
(all channels with the most significant byte set to 1). This would provide a maximum of 128 lanes,
which should suffice for most users.

Nodes would also need to agree on the channel assignments during the P2P handshake. Currently, one
of the conditions for the handshake to succeed is that there must exist an intersection of P2P
channels. Since lanes are defined in the application logic, the nodes only need to agree on the
application version, as it already happens in the current implementation.

> ### Duality lane/priority

Reviewer 2: To confirm, now lineID == priority? And the higher the greater?

Author: That's right.

Reviewer 2: I assume this is written somewhere...

   governance proposals would be required to pass to update the lane definitions. A first proposal
   would be required for upgrading the application, because the lane classification logic (thus the
   application's code) needs to know the lane configuration beforehand. And a second proposal would
   be needed for upgrading the lanes via `ConsensusParams`.
   While it is true that SDK applications could pass a governance proposal with both elements together,
   it would be something to _always_ do, and it is not clear what the situation would be for non-SDK applications.
  
   Also, it is not clear in which order the proposals should apply. The community should be careful
   not to break performance between the passing of both proposals. The `gov` module could be
   modified to allow the two changes to be shipped in the same gov proposal, but this does not seem
   a feasible solution.

Moreover, these two alternatives have a common problem which is how to deal with nodes that are
late, possibly having lane definitions that do not match with those of nodes at the latest heights.

## Consequences

> TODO

Reviewer 2: Remove?

### Positive

- Application developers will be able to better predict when transactions will be disseminated and
  processed to be included in a block. This has direct impact on block creation and transaction
  latency.
- The mempool will be able to offer Quality of Service guarantees, which does not exist in the
  current implementation. 

> - Lanes will preserve the "FIFO ordering of transactions" property within the same class (with a

Reviewer 2: This is neutral to me.

Author: Changed to neutral.

  best effort approach, as the current implementation).
- Applications that are unaware of this feature, and therefore not classifying transactions in `CheckTX`
  will observe the same behavior from the mempool as the current implementation.  

> ### Negative

Reviewer 2: On negative, again, are we discussing the starvation problem of low priority lanes. I refer to the `Reap()` part, not the broadcast part, which seems to be implemented using a starvation-free algorithm.

Author: The Reap section did not mention the starvation-free property. I corrected it.

> - The total FIFO ordering of transaction dissemination and block creation may be broken when

Reviewer 2: Total FIFO? Eheheh

Author: Not what I meant :stuck_out_tongue_closed_eyes:. I rephrased.

- `default_priority` is in `priorities` (the default lane is not necessarily the lane with the
  lowest priority), and
- the list `priorities` is empty if and only if `default_priority` is 0.

### Initialization

Currently, querying the app for `Info` happens during the handshake between CometBFT and the app,
during the node initialization, and only when state sync is not enabled. The `Handshaker` first
sends an `Info` request to fetch the app information, and then replays any stored block needed to
sync CometBFT with the app. The lane information is needed regardless of whether state sync is
enabled, so one option is to query the app information outside of the `Handshaker`.

In this proposed approach, updating the lane definitions will require a single governance proposal
for updating the software. CometBFT will not need to deal with dynamic lane changes: it will just
need to set up the lanes when starting up (whether afresh or in recovery mode).

Different nodes also need to agree on the lanes they use. When a node connects to a peer, they both
perform a handshake to agree on some basic information (see `DefaultNodeInfo`), including the

> version of the application they are executing. Since the application includes the lane definitions,

Author: Oops I assumed wrongly. I changed this paragraph proposing to "modify the P2P handshake process to validate that the application versions are compatible". Does this change break something?

   channel for disseminating transactions, and we will implement in the mempool the logic for
   selecting the order in which to send transactions.
- [[Reaping transactions for creating blocks](#reaping-transactions-for-creating-blocks)]
  Transactions will be reaped from higher-priority lanes first, preserving intra-lane FIFO ordering.
- [[Prioritization logic](#prioritization-logic)] For disseminating and reaping transactions, the
   scheduling algorithm should be prevent starvation of low-priority lanes.

### Lanes definition

The list of lanes and their corresponding priorities will be hardcoded in the application logic. A
priority is a value of type `uint32`, with 0 being a reserved value (see below). The application
also needs to define which of the lanes it defines is the default lane. 
also needs to define which of the lanes in the list it defines is the **default lane**. 

To obtain the lane information from the application, we need to extend the ABCI `Info` response to
include the following fields. These fields need to be filled by the application only in case it
wants to implement lanes.
```protobuf
message InfoResponse {
  ...
  repeated uint32 lanes = 6;

>   uint32 default_lane = 7;

Reviewer 2: Is it correct to set the default lane to 7 if it is not in the list of lanes?

Author: `6` and `7` are not default values. They are the field number. Currently `InfoResponse` has 5 fields.

The lowest priority a lane may have is 1. The value 0 is reserved for two cases: when the
application does not classify the transaction (i.e. no priority returned) and for `CheckTx`
responses of invalid transactions.

On receiving the information from the app, CometBFT will validate that:
- `lane_priorities` has no duplicates (values in `lane_priorities` don't need to be sorted),

> - `default_lane_priority` is in `lane_priorities` (the default lane is not necessarily the lane with

Reviewer 2: This is violated in the example above.

Author: Same answer as above.

need to set up the lanes when starting up (whether afresh or in recovery mode).

Different nodes also need to agree on the lanes they use. When a node connects to a peer, they both
perform a handshake to agree on some basic information (see `DefaultNodeInfo`). Since the
application includes the lane definitions, it suffices to ensure that both nodes agree on the
version of the application. Although the application version is included in `DefaultNodeInfo`, there
is currently no check for compatibility between the versions. To address this, we would need to
modify the P2P handshake process to validate that the application versions are compatible.

Finally, this approach is compatible with applications that need to swap binaries when
catching up or upgrading, such as SDK applications using [Cosmovisor][cosmovisor].

> When a node is catching up (i.e. BlockSyncing), its peers will detect

Reviewer 2: ```suggestion
When a node is catching up (i.e., state or block syncing), its peers will detect
```

dissemination.

### Prioritization logic

For transaction dissemination and for reaping transactions for creating blocks we want a scheduling
algorithm that satisfies the properties "Priorities between classes" and "FIFO ordering per class".
This means that it must support selection by _weight_, ensuring each lane gets a fraction of the P2P
channel capacity proportional to its priority. Additionally, we want the algorithm to be _fair_ to
prevent starvation of low-priority lanes. 

A first option that meets these criteria is the current prioritization algorithm on the P2P reactor,
which we could easily reimplement in the mempool. It works as follows:
- On each P2P channel, the variable `recentlySent` keeps track of how many bytes were recently sent
  over the channel. Every time data is sent, increase `recentlySent` with the number of bytes
  written to the channel. Every 2 seconds, decrease `recentlySent` by 20% on all channels (these
  values are fixed).
- When sending the next message, [pick the channel][selectChannelToGossipOn] whose ratio
  `recentlySent/Priority` is the least.

From the extensive research in operating systems and networking, we can choose for the MVP an

> existing scheduling algorithm that meet these requirements and are straightforward to implement,

Reviewer: ```suggestion
existing scheduling algorithm that meets these requirements and are straightforward to implement,
```

> such as a variant of the [Weighted Round Robin][wrr] (WRR) algorithm. Another one to consider is
> [Weighted Random Early Detection][wred].

Reviewer: ```suggestion
such as a variant of the [Weighted Round Robin][wrr] (WRR) algorithm.
```

I was wrong w.r.t. WRED, it's rather used to help congestion control algorithms (in the presence of QoS), so not really related here.

- Cosmos SDK's [gas prices][sdk-gas-prices]
- Cosmos SDK's [application-side mempool][sdk-app-mempool]
- Skip's [Block SDK][skip-block-sdk]
- P2P's [selectChannelToGossipOn][selectChannelToGossipOn] function
- [Weighted Round Robin][wrr]

> - [Weighted Random Early Detection][wred]

Reviewer: ```suggestion
```

[sdk-gas-prices]: https://docs.cosmos.network/v0.50/learn/beginner/tx-lifecycle#gas-and-fees
[sdk-app-mempool]: https://docs.cosmos.network/v0.47/build/building-apps/app-mempool
[skip-block-sdk]: https://github.com/skip-mev/block-sdk/blob/v2.1.3/README.md
[cosmovisor]: https://docs.cosmos.network/v0.50/build/tooling/cosmovisor
[selectChannelToGossipOn]: https://github.com/cometbft/cometbft/blob/6d3ff343c2d5a06e7522344d1a4e17d24ce982ad/p2p/conn/connection.go#L542-L563
[wrr]: https://en.wikipedia.org/wiki/Weighted_round_robin

> [wred]: https://en.wikipedia.org/wiki/Weighted_random_early_detection

Reviewer: ```suggestion
```
